Integrate `configManager` in `src/index/IndexImpl.cpp`

schlegan opened the pull request:
I replaced as much of the old parsing of configuration option with the new `ConfigManager` as possible, in order to make this whole thing easier and easier to read.

schlegan committed:
src/index/IndexImpl: Replaced manuell reading of a file to json with the fileToJson helper function.

schlegan committed:
src/util/ConfigManager/ConfigManager: Had to make the getConfigOption function public. Otherwise I can't check the set state of ConfigOptions.

schlegan committed:
src/index/IndexImpl: Tried to integrate ConfigManager with readIndexBuilderSettingsFromFile.

schlegan committed:
src/index/Vocabulary: Added missing explicit template instantiation.

schlegan committed:
src/util: Fixed cyclic dependency.

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

schlegan committed:
src/index/IndexImpl: Tried to integrate the ConfigManager in readConfiguration.

schlegan committed:
src/util/ConfigManager/ConfigManager: Renamed createConfigOption to addOption.

schlegan committed:
src/util/ConfigManager/ConfigManager: Replaced the choosing of default value existence with multiple function overloads.

schlegan committed:
src/util/ConfigManager/ConfigManager: Function for creating ConfigOptions now return a pointer to the created ConfigOption.

schlegan committed:
Cleand up some code formatting.

schlegan committed:
Merge branch 'master' into ConfigManagerUsageImprovements

schlegan committed:
Cleaned up some code formatting.

schlegan committed:
Merge branch 'ConfigManagerUsageImprovements' into IntegrateConfigManagerInIndexImpl

schlegan committed:
src/index/IndexImpl: Added the correct default values to the configOptions.

schlegan committed:
Merge branch 'master' into ConfigManagerUsageImprovements

schlegan committed:
Cleaned up some code formatting.

schlegan committed:
Merge branch 'ConfigManagerUsageImprovements' into IntegrateConfigManagerInIndexImpl

schlegan committed:
test/ConfigManagerTest: Shortend test code.

schlegan committed:
Merge branch 'ConfigManagerUsageImprovements' into IntegrateConfigManagerInIndexImpl

schlegan committed:
src/index/IndexImpl: Shortend code for the creation of configOptions.

schlegan committed:
benchmark/BenchmarkExamples: Made the usage of addOption clearer.

schlegan committed:
src/util/ConfigManager/ConfigManager: AddOption now returns a reference instead of a pointer.

schlegan committed:
src/util/ConfigManager/ConfigManager: Improved element access in implementation by usage of std::views.

schlegan committed:
src/util/ConfigManager/ConfigManager: Improved consistency of examples for addOption.

schlegan committed:
Cleaned up some code formatting.

schlegan committed:
Merge branch 'ConfigManagerUsageImprovements' into IntegrateConfigManagerInIndexImpl

schlegan committed:
src/index/IndexImpl: Adjusted to changes in API.

schlegan committed:
src/index/IndexImpl: Deleted uneeded comments.

schlegan committed:
Cleaned up code formatting.

schlegan committed:
src/util/ConfigManager/ConfigManager: Replaced specific hash map with ad_utility:hash_map.

schlegan committed:
Merge branch 'ConfigManagerUsageImprovements' into IntegrateConfigManagerInIndexImpl

schlegan committed:
src/util/ConfigManager/ConfigManager: Improved helper function name.

schlegan committed:
src/util/ConfigManager/ConfigManager: Got rid of uneeded forwarding.

schlegan committed:
Merge branch 'ConfigManagerUsageImprovements' into IntegrateConfigManagerInIndexImpl

schlegan committed:
src/util/ConfigManager/ConfigManager: Fixed a missspelling from a previous function renaming.

joka921 commented on src/index/IndexImpl.cpp:
```diff
@@ -783,56 +789,105 @@ void IndexImpl::writeConfiguration() const {
+                   "The date of the last breaking change of the index format.",
+                   &indexFormatVersionDate);
+
+  // One of the options can't be done with `ConfigManager`.
```
Which one is that?
And then you can still use `fileToJson one and then simply parse the `configurationJson_`.

schlegan replied:
Can't remember, but I'm pretty sure, this comment is false anyway.

schlegan replied:
Done, see [eaecacc](https://github.com/ad-freiburg/qlever/pull/1041/commits/eaecacce58ecdf0424ab77ee976011c6c23320d9) and [e87e61c](https://github.com/ad-freiburg/qlever/pull/1041/commits/e87e61c5133955df3babf81fe286dfaaa7aedf04).

joka921 commented on src/index/IndexImpl.cpp:
```diff
@@ -783,56 +789,105 @@ void IndexImpl::writeConfiguration() const {
+  config.parseConfig(fileToJson(onDiskBase_ + CONFIGURATION_FILE));
+
+  if (!gitHash.empty()) {
+    configurationJson_["git-hash"] = gitHash;
```
```suggestion
```

schlegan replied:
Done, see [eaecacc](https://github.com/ad-freiburg/qlever/pull/1041/commits/eaecacce58ecdf0424ab77ee976011c6c23320d9).

joka921 reviewed (commented):
Some initial comments.

schlegan committed:
src/util/ConfigManager/ConfigManager: Sub managers can now be created for better organization.

schlegan committed:
Cleaned up code formatting.

schlegan committed:
Merge branch 'master' into ConfigManagerUsageImprovements

schlegan committed:
Merge branch 'ConfigManagerUsageImprovements' into IntegrateConfigManagerInIndexImpl

schlegan committed:
src/index/IndexImpl: No longer write to jsonConfiguration in readConfiguration.

schlegan commented on src/index/IndexImpl.cpp:
```diff
@@ -844,59 +899,28 @@ void IndexImpl::readConfiguration() {
-  }
+  configurationJson_["locale"]["language"] = lang;
+  configurationJson_["locale"]["country"] = country;
+  configurationJson_["locale"]["ignore-punctuation"] = ignorePunctuation;
```
Done in [eaecacc](https://github.com/ad-freiburg/qlever/pull/1041/commits/eaecacce58ecdf0424ab77ee976011c6c23320d9).

Are you sure about that? I distinctly remember, that `readConfiguration` was already writing to `configurationJson_` before I started this PR.

schlegan committed:
src/index/IndexImpl: Improved setting of configurationJson_ in readConfiguration via helper function.

schlegan commented on src/index/IndexImpl.cpp:
```diff
@@ -783,56 +789,105 @@ void IndexImpl::writeConfiguration() const {
-      throw std::runtime_error{
-          "Incompatible index format, see log message for details"};
+  // Is the index format version up to date?
+  auto indexFormatVersion = qlever::IndexFormatVersion{
+      indexFormatVersionPullRequestNumber,
+      DateOrLargeYear::parseXsdDate(indexFormatVersionDate)};
```
You can now add sub managers. We will discuss the rest later.

schlegan committed:
src/index/IndexImpl: Got rid of uneeded wasSetAtRuntime()-check.

schlegan committed:
src/index/IndexImpl: Local config manager now directly writes to numTriplesPerBatch_ and parserBatchSize_.

schlegan committed:
src/index/IndexImpl: Index version will now be checked before parsing.

schlegan committed:
src/index/IndexImpl: Fixed mistake, where the values needed for the index format version weren't registered in the config manager.

schlegan committed:
Cleaned up code formatting.

schlegan committed:
Github was being buggy.

schlegan commented on src/index/IndexImpl.cpp:
```diff
@@ -783,56 +789,105 @@ void IndexImpl::writeConfiguration() const {
-    if (configurationJson_["prefixes"]) {
+  if (prefixesOption.wasSetAtRuntime()) {
+    configurationJson_["prefixes"] = boolPrefixes;
+    if (boolPrefixes) {
```
Done.

schlegan commented on src/index/IndexImpl.cpp:
```diff
@@ -966,71 +1033,47 @@ void IndexImpl::readIndexBuilderSettingsFromFile() {
-              << " This might influence performance during index build."
-              << std::endl;
-  }
+  parserBatchSize_ = parserBatchSize;
```
Done, see [da5584a](https://github.com/ad-freiburg/qlever/pull/1041/commits/da5584a39ddb10994584a9254e2ddfb6d25fe810).

schlegan commented on src/index/IndexImpl.cpp:
```diff
@@ -783,56 +789,105 @@ void IndexImpl::writeConfiguration() const {
+    if (indexFormatVersion.date_.toBits() > currentVersion.date_.toBits()) {
+      LOG(ERROR) << "The version of QLever you are using is too old for this "
+                    "index. Please use a version of QLever that is "
+                    "compatible with this index"
```
Done, see [a022ccb](https://github.com/ad-freiburg/qlever/pull/1041/commits/a022ccb992dc9f8bf283c2ce626972dbd9d6579e).

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

schlegan committed:
src/index/IndexImpl: Replaced the manual check for the states of the options under cli argument locale with a validator.

schlegan committed:
src/index/IndexImpl: Added TODO for possible validator candidate.

schlegan committed:
Cleaned up code formatting.

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

schlegan committed:
src/index/IndexImpl: Added sub manager for easier organization.

schlegan committed:
Cleaned up code formatting.

joka921 commented on src/index/IndexImpl.cpp:
```diff
@@ -966,112 +1050,74 @@ void IndexImpl::readIndexBuilderSettingsFromFile() {
+    turtleParserIntegerOverflowBehavior_ =
+        TurtleParserIntegerOverflowBehavior::OverflowingToDouble;
+  } else {
+    // TODO This can maybe be replaced with a validator, if the logging of the
+    // information is not needed.
+    AD_CONTRACT_CHECK(std::find(allModes.begin(), allModes.end(),
+                                parserIntegerOverflowBehavior) ==
+                      allModes.end());
+    LOG(ERROR) << "Invalid value for parser-integer-overflow-behavior"
+               << std::endl;
+    LOG(INFO) << "The currently supported values are "
+              << absl::StrJoin(allModes, ",") << std::endl;
```
I like the idea of a validator.
But what I like even more is to handle the conversion from the String to the enum directly in the config manager, s.t. we can for example call
````
configManager.addOptionWithConversion<string, someEnum>(path, description, &enum, possiblyThrowingConverterFunctionFromStringToEnum).


````

You can maybe already think of a way to incorporate this into the configManager.

schlegan replied:
Hmm, that is a tough one. However, I'm not so sure, if extending the `addOption` family would be the best idea in terms of clarity. Having 8 options for adding an option, with half being radically different to the other half, does sound a bit confusing.

I'm also not sure, if we can get around the need for explicitly stating the needed types in `<>`. It would be kinda weird, to have to pass variables for intermediate results.

My first thought, on how to accomplish this while converter idea, would have been to use the same style as for `addValidator`, but with having to pass a variable address at the beginning, to put the result in.
Of course, that would still require the user to personally build the variables for the intermediate results, so I'm not the biggest fan.

joka921 reviewed (commented):
This is already very nice and clean now.
I hope I soon will have the time to try it out and write some descriptions for this.
I only have two very abstract comments about which we maybe want to talk in person at some point.

schlegan commented on src/index/IndexImpl.cpp:
```diff
@@ -783,18 +789,74 @@ void IndexImpl::writeConfiguration() const {
+
+  // TODO Write a description.
+  std::vector<std::string> languagesInternal;
+  config.addOption("languages-internal", "", &languagesInternal, {"en"});
```
> I think in this function, There should be (almost) no default, I am not sure about the git hash though, but we'll find out.

Quite possible. However, I got most of the default values directly from the old code, or from you, if I remember correctly. So I'm not `100%` sure.

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

schlegan committed:
Fixed code formatting.

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

schlegan committed:
src/index/IndexImpl.cpp: Added description to the validator.

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

joka921 committed:
Added some comments.

joka921 committed:
Added some preliminary descriptions to the config manager in IndexImpl.cpp

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

schlegan commented:
The newest merge with the `master` branch contained some default values for options, that didn't have default values before. So I've added them:

```
config.addOption("num-subjects-normal", "", &numSubjectsNormal_, 0UL);
config.addOption("num-objects-normal", "", &numObjectsNormal_, 0UL);
```

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

joka921 committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

# Conflicts:
#	src/index/IndexImpl.cpp

joka921 committed:
Merge in the master, and see for myself what is going on there.

joka921 committed:
Merge remote-tracking branch 'andre/IntegrateConfigManagerInIndexImpl' into IntegrateConfigManagerInIndexImpl

# Conflicts:
#	src/index/IndexImpl.cpp

schlegan committed:
src/util/ConfigManager/ConfigManager: Wrote two public helper functions for printConfigurationDoc, so that user have more control over the representation.

schlegan committed:
src/index/IndexImpl::readIndexBuilderSettingsFromFile: Moved the creation of the config manager into its own file.

schlegan committed:
src/index/IndexBuilderMain: The '-h' cli parameter now also prints the configuration options for the index builder.

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

joka921 commented on src/index/IndexImpl.cpp:
```diff
@@ -1083,11 +1089,13 @@ void IndexImpl::readIndexBuilderSettingsFromFile() {
-  LOG(INFO) << "You specified \"locale = " << lang << "_" << country << "\" "
-            << "and \"ignore-punctuation = " << ignorePunctuation << "\""
-            << std::endl;
+  LOG(INFO) << "You specified \"locale = " << configVariables.localLang_ << "_"
+            << configVariables.localCountry_ << "\" "
+            << "and \"ignore-punctuation = "
+            << configVariables.localIgnorePunctuation_ << "\"" << std::endl;
```
I think we can delete those lines now that we have the automatic printing of the configuration.

schlegan replied:
Done, see [76b5747](https://github.com/ad-freiburg/qlever/pull/1041/commits/76b574703a9a7fe7f260e5a4c0ebcdc9544f710f).

joka921 commented on src/index/IndexImpl.h:
```diff
@@ -631,6 +632,42 @@ class IndexImpl {
+  // their string representation.
+  inline static const ad_utility::HashMap<std::string,
+                                          TurtleParserIntegerOverflowBehavior>
+      turtleParserIntegerOverflowBehaviorMap_{
```
This can be more readable if you assign it by an immediately invoked lambda inside ow which you use `using enum TurtleParserIntegerOverflowBehaviior`.

schlegan replied:
Done, see [c366a3f](https://github.com/ad-freiburg/qlever/pull/1041/commits/c366a3f1f42f3f28e058c7ba60a7fdc3ae029548).

joka921 commented on src/index/Index.h:
```diff
@@ -281,4 +281,9 @@ class Index {
+  /*
+  @brief Print the detailed documentation of the options for the index builder.
+  */
+  std::string getConfigurationDocForIndexBuilder();
```
I think this should return also two strings: the JSON and the detailed configuration separately, then we can configure the printing in `IndexBuilderMain` (I will think of something).

schlegan replied:
Shouldn't the configuration of the printing happen inside `IndexImpl::getConfigurationDocForIndexBuilder`? Separation of concerns and all that.

joka921 reviewed (changes requested):
This is much much better already. I think we are almost there.

joka921 committed:
Refactor workflows + revert to standard `apt` (#1185)

1. For a while now, we used `apt-spy` because `apt` was not working reliably within GitHub's workflows. Recently, it was `apt-spy` that was unreliable and so we are reverting back to `apt`
2. Use the occasion to factor out separate `action.yml` "sub"-workflow for installing a compiler on Ubuntu and for installing dependencies on Ubuntu, which are used at various places in our workflows

joka921 committed:
Make sonarcloud happy by making an `std::move` explicit (#1179)

schlegan committed:
Benchmark::ResultTable : Add helper functions for column-based operations (#1174)

Add some helper functions for adding statistics columns (that is columns with results that were not measured but computed from other columns) to the result table of a benchmark run.

kcaliban committed:
For GROUP BY optimization, use vector offsets to access aggregation data. (#1181)

In the hash map-based GROUP BY optimization, the hash map now only stores a mapping from ID->size_t where the size_t is an index into a vector that stores the actual data. This optimization will make the support for multiple different aggregates much more efficient.

joka921 committed:
Push to Docker Hub as `adfreiburg/qlever` (#1192)

So far, the Docker images were pushed to Docker Hub as `<private username>/qlever`, to make it easier to check whether everything works as it should. Everything seems to work as it should and so now we push as the official `adfreiburg/qlever`.

joka921 committed:
Make sure that no exceptions get lost in the parallel Turtle parser (#1187)

So far, the parallel Turtle parser had the problem that certain exceptions disappeared. After #1124, two exceptions happened more often, namely that the `FILE_BUFFER_SIZE` was not large enough (it was reduced from 100M to 10M), and an overflow of the sort key buffer (its preallocation was increased significantly to save time). So far, both exceptions had the effect that the parsing hung forever, either because batches were still being produced but not processed, or batches were still being processed but not produced.

All exceptions are now caught and appear in the log. The potential overflow of the sort key buffer is now fixed. The `FILE_BUFFER_SIZE` can still be too small (and it happens for `UniProt` and `OSM Planet`), but now the exception shows in the log and ends the index build.

RobinTF committed:
Update CTRE and embed via FetchContent (#1199)

In the future we should manage more (or even all) of our git dependencies via FetchContent, as it is much simpler than working with git submodules.

schlegan committed:
src/index/IndexImpl.cpp: Deleted uneeded std::make_pair.

schlegan committed:
src/index/IndexImpl.cpp: Deleted manuell documentation, that was replicated by the generated documentation of ConfigManager.

schlegan commented on src/index/IndexImpl.h:
```diff
@@ -631,6 +632,42 @@ class IndexImpl {
 
+  // Assigns the entries of the enum `TurtleParserIntegerOverflowBehavior` to
+  // their string representation.
+  inline static const ad_utility::HashMap<std::string,
```
Ah, but only with `inline` can I use a default value and without `static` every instance of the class has its own copy, even though they are never different.

schlegan committed:
src/index/IndexImpl.h: Improved readability of the initialization value of turtleParserIntegerOverflowBehaviorMap_.

schlegan committed:
Fixed code formatting.

schlegan commented on src/index/IndexImpl.cpp:
```diff
@@ -1041,40 +1039,48 @@ void IndexImpl::readIndexBuilderSettingsFromFile() {
+              std::views::keys(turtleParserIntegerOverflowBehaviorMap_), ", "),
       "dummy description for the overflow behavior validator", overflowOption);
 
+  return std::make_pair(std::move(config), std::move(variables));
```
Done, see [2a55e50](https://github.com/ad-freiburg/qlever/pull/1041/commits/2a55e502e5cf1f326f7d0c85b9001930c00b3df5).

schlegan committed:
src/index/IndexImpl.cpp: Got rid of uneeded explicit lambda return type.

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

schlegan committed:
src/util/ConfigManager/ConfigManager: Git somehow forgot, that I've added this file.

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl

schlegan committed:
Merge branch 'master' into IntegrateConfigManagerInIndexImpl
